src/ArrowFunctions.cpp

```diff
@@ -617,18 +615,29 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize) // adjust batchSize if needed
               batchSize = numElems - i;
-            // unlike the base case, the list stores NaN values and is able to read them out. 
-            // This prevents the need for any additional processing
-            // setting nullptr for def and rep levels allows use to ignore the empty segments since we only care about values here.
-            // However, for floats we still need a temporary array to allow for conversion to double
-            float* tmpArr = new float[batchSize] { 0 }; // this will not include NaN values
-            (void)reader->ReadBatch(batchSize, nullptr, nullptr, tmpArr, &values_read);
-
-            for(int64_t j = 0; j < values_read; j++){
-              chpl_ptr[i+j] = (double)tmpArr[j];
+            float* tmpArr = new float[batchSize]; // this will not include NaN values
+            int16_t* def_lvl = new int16_t[batchSize];
+            int16_t* rep_lvl = new int16_t[batchSize];
+            (void)reader->ReadBatch(batchSize, def_lvl, rep_lvl, tmpArr, &values_read);
+
+            int64_t tmp_offset = 0; // used to properly access tmp after NaN encountered
+            int64_t val_ct = 0;
+            int64_t j = 0;
+            while (val_ct < values_read){
```

Comment:
I think this is the cause of the edge case i noted in the tests (all NAN's after `values_read` are treated as zeroes) because we exit the while loop after the last non-NAN. Maybe we intialize the tmpArr to be `NAN`s? Or keep looping until we've hit everything in `def_lvl` and keep writing `NAN`s when it equals 2? idk if that would fix the problem, i'm just spitballing at this point

Reply:
This was definitely on the right track. I made a few adjustments to keep things as simple as possible.